Gems/Terrain/Code/Source/TerrainRenderer/TerrainFeatureProcessor.cpp

```diff
@@ -99,65 +100,146 @@ namespace Terrain
     }
 
     void TerrainFeatureProcessor::Deactivate()
     {
+        AzFramework::Terrain::TerrainDataNotificationBus::Handler::BusDisconnect();
+        AZ::RPI::MaterialReloadNotificationBus::Handler::BusDisconnect();
+
         m_patchModel = {};
         m_areaData = {};
-        AZ::RPI::MaterialReloadNotificationBus::Handler::BusDisconnect();
     }
 
     void TerrainFeatureProcessor::Render(const AZ::RPI::FeatureProcessor::RenderPacket& packet)
     {
         ProcessSurfaces(packet);
     }
 
-    void TerrainFeatureProcessor::UpdateTerrainData(
-        const AZ::Transform& transform,
-        const AZ::Aabb& worldBounds,
-        float sampleSpacing,
-        uint32_t width, uint32_t height, const AZStd::vector<float>& heightData)
+    void TerrainFeatureProcessor::OnTerrainDataDestroyBegin()
+    {
+        m_areaData = {};
+    }
+    
+    void TerrainFeatureProcessor::OnTerrainDataChanged(const AZ::Aabb& dirtyRegion, TerrainDataChangedMask dataChangedMask)
     {
-        if (!worldBounds.IsValid())
+        if (dataChangedMask != TerrainDataChangedMask::HeightData && dataChangedMask != TerrainDataChangedMask::Settings)
         {
             return;
         }
 
+        AZ::Aabb worldBounds = AZ::Aabb::CreateNull();
+        AzFramework::Terrain::TerrainDataRequestBus::BroadcastResult(
+            worldBounds, &AzFramework::Terrain::TerrainDataRequests::GetTerrainAabb);
+
+        const AZ::Aabb& regionToUpdate = dirtyRegion.IsValid() ? dirtyRegion : worldBounds;
+
+        m_dirtyRegion.AddAabb(regionToUpdate);
+        m_dirtyRegion.Clamp(worldBounds);
+
+        AZ::Transform transform = AZ::Transform::CreateTranslation(worldBounds.GetCenter());
+        
+        AZ::Vector2 queryResolution = AZ::Vector2(1.0f);
+        AzFramework::Terrain::TerrainDataRequestBus::BroadcastResult(
+            queryResolution, &AzFramework::Terrain::TerrainDataRequests::GetTerrainHeightQueryResolution);
+
+        uint32_t width = aznumeric_cast<uint32_t>((float)m_dirtyRegion.GetXExtent() / queryResolution.GetX());
+        uint32_t height = aznumeric_cast<uint32_t>((float)m_dirtyRegion.GetYExtent() / queryResolution.GetY());
+        
+        uint32_t worldWidth = aznumeric_cast<uint32_t>((float)worldBounds.GetXExtent() / queryResolution.GetX());
+        uint32_t worldHeight = aznumeric_cast<uint32_t>((float)worldBounds.GetYExtent() / queryResolution.GetY());
+
         m_areaData.m_transform = transform;
         m_areaData.m_heightScale = worldBounds.GetZExtent();
         m_areaData.m_terrainBounds = worldBounds;
-        m_areaData.m_heightmapImageHeight = height;
-        m_areaData.m_heightmapImageWidth = width;
-        m_areaData.m_sampleSpacing = sampleSpacing;
+        m_areaData.m_heightmapImageWidth = worldWidth;
+        m_areaData.m_heightmapImageHeight = worldHeight;
+        m_areaData.m_updateWidth = width;
+        m_areaData.m_updateHeight = height;
+        m_areaData.m_sampleSpacing = queryResolution.GetX();
```

Comment:
Should add a comment explaining that different query resolutions in X and Y aren't supported (right now?).

Reply:
Actually can we drop multiple-dimension query resolution from the terrain system itself? why would anyone not want this to be uniform. Unlikely rendering or physics would ever respect different values per dimension.